### ISSUE TITLE
Provide Expr.to_json()

### ISSUE THREAD

**Issue by PaulRudin** [pola-rs/polars#8482]
### Problem description

I want to be able to use polars to produce json directly from data in a dataframe. Most (all?) polars data types have an obvious representation as json; and providing a mechanism to do this as part of a polars expression potentially allows for parallelizing obtaining json from data in a dataframe and in any case (at least in the case of python) should be a lot faster than using apply.

**Comment by zundertj** [comment]
We can currently turn a dataframe into json by doing
```python
df_as_json = df.write_json(file=None)
```
How would this look for individual expressions/columns? As the JSON is represents the whole dataframe, not a single record. I think an example of input and output would be useful here.

**Comment by PaulRudin** [comment]
Every(?) datatype has a natural representation as json, so as an expression you just get the string that rendering it as json normally gives. e.g.

So if you have a column "a" of type int32, then `df.with_columns("a").to_json()` would give the string representation of those numbers, and so on for other datatypes. If there's already code to serialise a whole dataframe as json, it's probably not too hard to do what I have in mind. I'll take a look at it...

**Comment by thomasaarholt** [comment]
You can now serialize and deserialize expressions as:

```python
import polars as pl

expr1 = 2*pl.col("foo") + 1
json_repr = expr1.meta.write_json()

expr2 = pl.Expr.from_json(json_repr)


print(expr1) # [([(2) * (col("foo"))]) + (1)]
print(expr2) # [([(2) * (col("foo"))]) + (1)]
```
I'd show the above with an `assert`, but you can't compare expressions with `==` since that would create a new expression

**Comment by cmdlineluser** [comment]
I think this is about serializing the actual values @thomasaarholt 

i.e. [`.struct.json_encode()`](https://pola-rs.github.io/polars/py-polars/html/reference/expressions/api/polars.Expr.struct.json_encode.html#polars.Expr.struct.json_encode) except not restricted to structs.

So that the following could also work:

```python
pl.lit([1, 2, 3]).json_encode() # "[1, 2, 3]"
```

**Comment by thomasaarholt** [comment]
Oh! My mistake! An example would make it a bit more obvious.

So something that would prevent having to jump through this hoop?

I think the above response is good though - what should the row content look like? I think valid json requires a key.

```python
df = pl.DataFrame(
    {
        "foo": [[1,2,3], [3,2,1]],
    }
)
df.with_columns(json_repr = pl.struct(pl.col("foo")).struct.json_encode())

# shape: (2, 2)
# ┌───────────┬─────────────────┐
# │ foo       ┆ json_repr       │
# │ ---       ┆ ---             │
# │ list[i64] ┆ str             │
# ╞═══════════╪═════════════════╡
# │ [1, 2, 3] ┆ {"foo":[1,2,3]} │
# │ [3, 2, 1] ┆ {"foo":[3,2,1]} │
# └───────────┴─────────────────┘
```

**Comment by thomasaarholt** [comment]
Perhaps a string representation of any dtype would be enough, although it currently doesn't work:

```python
df = pl.DataFrame(
    {
        "foo": [[1,2,3], [3,2,1]],
    }
)
df.select(json_repr = pl.col("foo").cast(pl.Utf8)) # this won't run due to the pl.List dtype

df_expected = pl.DataFrame({"json_repr": ['[1, 2, 3]', '[3, 2, 1]']})
```

**Comment by cmdlineluser** [comment]
Yes is in that case, `df_expected` would be correct.

It's probably easier to see with a nested list example:

```python
df = pl.DataFrame({"foo": [[dict(a=1, b=2)]]})

df.with_columns(json_repr = pl.col("foo").map_elements(lambda x: json.dumps(list(x))))

# shape: (1, 2)
# ┌─────────────────┬────────────────────┐
# │ foo             ┆ json_repr          │
# │ ---             ┆ ---                │
# │ list[struct[2]] ┆ str                │
# ╞═════════════════╪════════════════════╡
# │ [{1,2}]         ┆ [{"a": 1, "b": 2}] │
# └─────────────────┴────────────────────┘
```
It was described as wanting an equivalent to `.map_elements(json.dumps)`

It was added in #11728 but only for structs.

**Comment by DeflateAwning** [comment]
I propose this can be closed in favour of https://github.com/pola-rs/polars/issues/14029, which looks like it's pretty close to being completed.

**Comment by jjurm** [comment]
This is still a very relevant feature request.

There was a prior attempt to add root-level JSON encoding via `pl.Expr.json_encode()` in https://github.com/pola-rs/polars/pull/18396, but sadly it wasn’t finished/merged.

From a user/API perspective, this functionality really belongs to **`pl.Expr` (root namespace) for *all dtypes*** (struct/list/primitive, and arbitrarily nested combinations). The fact that users can already *emulate* “JSON-encode any expression” by wrapping it into a temporary struct strongly suggests that a native `Expr.json_encode()` is the right abstraction (taken from [this comment](https://github.com/pola-rs/polars/issues/14029#issuecomment-3526365743)):

```python
def json_encode(expr: pl.Expr) -> pl.Expr:
    return pl.struct([expr.alias("")]).struct.json_encode().str.json_path_match("$['']")

print(df.select(pl.col("any_column").pipe(json_encode)))
```

---

_EDIT: the above actually doesn't work for non-nested strings: when converted to json, strings should be escaped and quoted in `""`, while the code above outputs them verbatim. Instead, one can use the following workaround:_

```python
def json_encode(expr: pl.Expr) -> pl.Expr:
    return pl.struct([expr.alias("")]).struct.json_encode().str.strip_prefix('{"":').str.strip_suffix('}')
```

_This still has drawbacks: fails if the passed `expr` references multiple columns, and doesn't preserve the expression name._

---

If/when root-level `Expr.json_encode()` exists (for all dtypes), it should supersede the namespace-specific approaches:

- #14029 would become obsolete, because lists are just one of many dtypes that should be JSON-encodable via `pl.Expr.json_encode()`.
- #19407 could then be closed, similarly to how **#18353** (list-only JSON encoding) was closed, in favor of a more universal API.

In short: finishing and shipping `pl.Expr.json_encode()` (as in the direction of #18396) as a type-agnostic operation would provide a consistent, composable solution and remove the need for dtype-specific JSON encoding methods.
